### ISSUE TITLE
npm@3: `npm install -S` reverts shrinkwrap with devDependencies

### ISSUE THREAD

**Issue by wbinnssmith** [npm/npm#9647]
Hi all! Thanks for npm and congratulations on the stable release of npm@3 :) Apologies if this issue is known; I couldn't find anything similar.

I absolutely love the fact that `npm install --save` will update an existing shrinkwrap file if it's there. Unfortunately, it looks like if that shrinkwrap file was generated with `npm shrinkwrap --dev`, npm will re-write the shrinkwrap file with just the production dependencies (and often output a ton of WARN messages that it's excluding devDependencies).

What's interesting is that the shrinkwrap I generated *initially* with `npm shrinkwrap --dev` includes the devDependencies, but they're actually listed under the `dependencies` key. Maybe this has something to do with it — I imagine npm would/could use the presence of a `devDependencies` key as a hint to continue generating a shrinkwrap with devDependencies on events like `npm install --save`.

**tl;dr reproduction steps on npm@3.3.3 and 3.3.4 on a project with both dependencies and devDependencies:**

1. `npm shrinkwrap --dev`
2. the generated shrinkwrap will contain devDependencies inside a "dependencies" key
2. `npm i -S xtend` (or any other module)
3. npm will install the module, but will begin shrinkwrapping without dev dependencies and issue WARNs
4. the resulting shrinkwrap won't include devDependencies that were present after step 1

Thanks and congratulations again!

**Comment by othiym23** [comment]
Yeah, this looks like an edge case we weren't catching. We should probably have some way of checking whether `devDependencies` were included in `npm-shrinkwrap.json`, and preserve them when upgrading the shrinkwrap file. It would also be nice to do this without adding any additional metadata to `npm-shrinkwrap.json` (which might break backwards compatibility for the file), but we'll see how it goes. @iarna, what do you think?

**Comment by wbinnssmith** [comment]
I don't mind tackling this myself, actually, maybe with guidance from @iarna? I can't seem to find any test cases around saving the shrinkwrap post-save, so I'll probably PR that first and we'll go from there. What do you think?

**Comment by micahr** [comment]
This bug introduced some very unexpected behavior when installing new dependencies (namely deleting lots of lines from our shrinkwrap file). Is there any work around while the bug is being worked on, short of going back to NPM 2?

**Comment by iarna** [comment]
This PR didn't make it under last week's release cut off, so it'll be included in this week's release instead.
